sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/QueueingBeamFnDataClient.java

```diff
@@ -193,16 +287,24 @@ public void accept(T value) throws Exception {
     }
   }
 
-  static class ConsumerAndData<T> {
-    private final FnDataReceiver<T> consumer;
-    private final T data;
+  private static class ConsumerAndData<T> {
+    private final @Nullable FnDataReceiver<T> consumer;
+    private final @Nullable T data;
+
+    public ConsumerAndData() {
+      this.consumer = null;
+      this.data = null;
+    }
 
     public ConsumerAndData(FnDataReceiver<T> receiver, T data) {
       this.consumer = receiver;
       this.data = data;
     }
 
     void accept() throws Exception {
+      if (consumer == null || data == null) {
+        throw new IllegalStateException("cannot accept POISON");
```

Comment:
As this is dealing with a special case, can you leave some comments to explain when this could happen? (e.g. consumer or data is null then it must be POSION)

Reply:
```
        ConsumerAndData<?> tuple = queue.take();
        if (tuple == null) {
          break; // queue has been drained and is closed.
        }
```
in the drainAndBlock code I am seeing this. So will POSION  be a ConsumerAndData that has null consumer or null data.

Reply:
I modified the poision so it is more self-contained, it just implements a reciever that throws.  Now ConsumerAndData always has non-null elements (enforced by nullness checking and default of NonNull annotation) and the poison stuff is contained entirely with in ClosableQueue.

The user of the queue just knows that take() returns null once the queue is empty and has been closed, otherwiise non-null and never the poison.